shadow_sampler.py

```diff
@@ -0,0 +1,104 @@
+import os.path as osp
+
+import argparse
+import torch
+import torch.nn.functional as F
+from torch_geometric.datasets import Flickr
+from torch_geometric.loader import GraphSAINTRandomWalkSampler
+from torch_geometric.loader import ShaDowKHopSampler
+from torch_geometric.nn import GraphConv
+from torch_geometric.utils import degree
+
+path = osp.join(osp.dirname(osp.realpath(__file__)), '..', 'data', 'Flickr')
+dataset = Flickr(path)
+data = dataset[0]
+row, col = data.edge_index
+data.edge_weight = 1. / degree(col, data.num_nodes)[col]  # Norm by in-degree.
+
+parser = argparse.ArgumentParser()
+parser.add_argument('--use_normalization', action='store_true')
+args = parser.parse_args()
+
+
+indices = data.train_mask.nonzero()
+index = torch.flatten(indices)
+loader = ShaDowKHopSampler(data, depth=2, num_neighbors=5, node_idx=index)
+
+
+class Net(torch.nn.Module):
+    def __init__(self, hidden_channels):
+        super(Net, self).__init__()
+        in_channels = dataset.num_node_features
+        out_channels = dataset.num_classes
+        self.conv1 = GraphConv(in_channels, hidden_channels)
+        self.conv2 = GraphConv(hidden_channels, hidden_channels)
+        self.conv3 = GraphConv(hidden_channels, hidden_channels)
+        self.lin = torch.nn.Linear(3 * hidden_channels, out_channels)
+
+    def set_aggr(self, aggr):
+        self.conv1.aggr = aggr
+        self.conv2.aggr = aggr
+        self.conv3.aggr = aggr
+
+    def forward(self, x0, edge_index, edge_weight=None):
+        x1 = F.relu(self.conv1(x0, edge_index, edge_weight))
+        x1 = F.dropout(x1, p=0.2, training=self.training)
+        x2 = F.relu(self.conv2(x1, edge_index, edge_weight))
+        x2 = F.dropout(x2, p=0.2, training=self.training)
+        x3 = F.relu(self.conv3(x2, edge_index, edge_weight))
+        x3 = F.dropout(x3, p=0.2, training=self.training)
+        x = torch.cat([x1, x2, x3], dim=-1)
+        x = self.lin(x)
+        return x.log_softmax(dim=-1)
+
+
+device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
+model = Net(hidden_channels=256).to(device)
+optimizer = torch.optim.Adam(model.parameters(), lr=0.001)
+
+
+def train():
+    model.train()
+    model.set_aggr('add' if args.use_normalization else 'mean')
+
+    total_loss = total_examples = 0
+    for data in loader:
+        data = data.to(device)
+        optimizer.zero_grad()
+
+        if args.use_normalization:
+            edge_weight = data.edge_norm * data.edge_weight
+            out = model(data.x, data.edge_index, edge_weight)
```

Comment:
It seems the model is missing the pooling step to go from subgraph-level embeddings to node-level embeddings.

Reply:
I have added the pooling step.